fix(Core/Movement): Only update home position for patrolling NPCs

blinkysc opened the pull request:
[!] Please Read! This PR comes from AI with [MCP](https://github.com/blinkysc/azerothMCP)

## Changes Proposed:
This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

This PR conditionally calls `SetHomePosition(GetPosition())` only for patrolling NPCs, preventing stationary NPCs from drifting over time while allowing patrolling NPCs to resume from where they were interrupted.

**Root Cause:** When a player interacts with an NPC (gossip, quest, vendor), the handler would update the NPC's home position to their current position. Over thousands of interactions, small movements and terrain height variations would accumulate, causing stationary NPCs to drift 30+ units from their original spawn.

**The Fix:** Only call `SetHomePosition` for NPCs with `WAYPOINT_MOTION_TYPE` or currently on a waypoint path via SmartAI. This:
- Prevents drift for stationary NPCs (IDLE_MOTION_TYPE)
- Allows patrolling NPCs to resume from where they were interrupted

## Issues Addressed:
- Closes NPCs/questgivers drifting from their spawn locations over server uptime

## SOURCE:
The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick).

This is a fix for a regression introduced in PR #9272 (TrinityCore cherry-pick).

## Tests Performed:
This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [x] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

### STATIONARY NPCs (should NOT drift)
Use .gps after Dark Ranger or Kilix attacks as good tests. It should be many times (like 10-20 times)
- Dark Ranger Lyana `.go creature id 23778`
- Kilix the Unraveler `.go creature id 26653`
- Apothecary Lysander `.go creature id 24126`

Note: These stationary NPC's are the ones I've noticed most while questing.... Poor Kilix is at war in his little hole nonstop

### PATROLLING NPCs - DB Waypoints (should resume patrol after evade)
- Samantha Swifthoof `.go creature id 11748`
- Marcus Bel `.go creature id 5800`
- Thora Feathermoon `.go creature id 5798`
- Magister Aledis `.go creature id 20159`
     ^ 3 min timer between interactions
- Greishan Ironstove `.go creature id 3291`
- Lizzarik `.go creature id 3658`

### SMARTAI WAYPOINT NPCs (escort quests)
- The Defias Traitor `.go creature id 467`
- Miran `.go creature id 1379`
- Highlord Taelan Fordring `.go creature id 1842`
- Kinelory `.go creature id 2713`
- Therylune `.go creature id 3584`
- Deathstalker Adamant `.go creature id 3849`
- Sorcerer Ashcrombe `.go creature id 3850`
- Feero Ironhand `.go creature id 4484`
- Willix the Importer `.go creature id 4508`
- Blastmaster Emi Shortfuse `.go creature id 7998`
- Dorius Stonetender `.go creature id 8284`

## Known Issues and TODO List:

## How to Test AzerothCore PRs

When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

blinkysc committed:
fix(Core/Movement): Remove SetHomePosition calls causing NPC drift

Player interactions (gossip, quest, vendor) were calling
SetHomePosition(GetPosition()) which caused NPCs to gradually
drift from their spawn positions over time.

The PauseMovement() call is sufficient to stop NPCs during
interaction - updating home position is unnecessary and harmful.

Removed SetHomePosition from:
- NPCHandler.cpp (gossip interactions)
- QuestHandler.cpp (quest giver interactions)
- ItemHandler.cpp (vendor interactions)

Added unit tests to verify the fix and prevent regression.

Nyeriah commented:
Consider the scenario:

- NPC spawns in shattrath city 
- NPC starts pathing through entire Terokkar 
- Player talks to npc, npc stops moving 
- NPC engages a mob or for some other reason enters evade mode 

…

- NPC now walks back all the way to Shattrath city as opposed to where it stopped when talked to

blinkysc commented:
> Consider the scenario:
> 
> * NPC spawns in shattrath city
> * NPC starts pathing through entire Terokkar
> * Player talks to npc, npc stops moving
> * NPC engages a mob or for some other reason enters evade mode
> 
> …
> 
> * NPC now walks back all the way to Shattrath city as opposed to where it stopped when talked to

Yeah... That's a good consideration...

blinkysc committed:
fix(Core/Movement): Only update home position for patrolling NPCs

Prevents stationary NPCs from drifting over time by only calling
SetHomePosition for creatures with WAYPOINT_MOTION_TYPE or currently
on a waypoint path via SmartAI.

blinkysc commented:
Ok I think this is better approach for this drift

sudlud committed:
Merge branch 'master' into fix-npc-position-drift
